s2a/src/main/java/io/grpc/s2a/handshaker/S2AProtocolNegotiatorFactory.java

```diff
@@ -0,0 +1,249 @@
+import com.google.errorprone.annotations.ThreadSafe;
+import io.grpc.Channel;
+import io.grpc.internal.ObjectPool;
+import io.grpc.netty.GrpcHttp2ConnectionHandler;
+import io.grpc.netty.InternalProtocolNegotiator;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.InternalProtocolNegotiators;
+import io.grpc.netty.InternalProtocolNegotiators.ProtocolNegotiationHandler;
+import io.grpc.s2a.channel.S2AChannelPool;
+import io.grpc.s2a.channel.S2AGrpcChannelPool;
+import io.grpc.s2a.handshaker.S2AIdentity;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerAdapter;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.ssl.SslContext;
+import io.netty.util.AsciiString;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import javax.annotation.Nullable;
+
+/** Factory for performing negotiation of a secure channel using the S2A. */
+@ThreadSafe
+public final class S2AProtocolNegotiatorFactory {
+  @VisibleForTesting static final int DEFAULT_PORT = 443;
+  private static final AsciiString SCHEME = AsciiString.of("https");
+
+  /**
+   * Creates a {@code S2AProtocolNegotiatorFactory} configured for a client to establish secure
+   * connections using the S2A.
+   *
+   * @param localIdentity the identity of the client; if none is provided, the S2A will use the
+   *     client's default identity.
+   * @param s2aChannelPool a pool of shared channels that can be used to connect to the S2A.
+   * @return a factory for creating a client-side protocol negotiator.
+   */
+  public static InternalProtocolNegotiator.ClientFactory createClientFactory(
+      @Nullable S2AIdentity localIdentity, ObjectPool<Channel> s2aChannelPool) {
+    checkNotNull(s2aChannelPool, "S2A channel pool should not be null.");
+    S2AChannelPool channelPool = S2AGrpcChannelPool.create(s2aChannelPool);
+    return new S2AClientProtocolNegotiatorFactory(localIdentity, channelPool);
+  }
+
+  static final class S2AClientProtocolNegotiatorFactory
+      implements InternalProtocolNegotiator.ClientFactory {
+    private final @Nullable S2AIdentity localIdentity;
+    private final S2AChannelPool channelPool;
+
+    S2AClientProtocolNegotiatorFactory(
+        @Nullable S2AIdentity localIdentity, S2AChannelPool channelPool) {
+      this.localIdentity = localIdentity;
+      this.channelPool = channelPool;
+    }
+
+    @Override
+    public ProtocolNegotiator newNegotiator() {
+      return S2AProtocolNegotiator.createForClient(channelPool, localIdentity);
+    }
+
+    @Override
+    public int getDefaultPort() {
+      return DEFAULT_PORT;
+    }
+  }
+
+  /** Negotiates the TLS handshake using S2A. */
+  @VisibleForTesting
+  static final class S2AProtocolNegotiator implements ProtocolNegotiator {
+
+    private final S2AChannelPool channelPool;
```

Comment:
To avoid creating+recreating the Channel repeatedly, we can rely on the 1 second timer of SharedResourceHolder. But we have retry backoffs that start at 1 second, so that's probably too small. An approach we've done elsewhere is to have this ProtocolNegotiator hold an unused reference to the Channel. So in the constructor it grabs a reference and in `close()` it releases the reference. This keeps the Channel alive for the lifetime of this negotiator.

Since we don't actually want a connection that entire time, we then configure ManagedChannelBuilder.idleTimeout() with the amount of time it should wait before shutting down its connections and going to sleep.

On client-side, the ProtocolNegotiator lives for the life of all connections. So we could avoid each connection getting its own reference to the Channel. But on server-side connections can outlive the negotiator. So it still makes sense for the Handler to acquire its own reference from the pool. (Yes, there's no server-side now, but it seems we'd expect that to happen eventually.)

Reply:
Thank you for the explanation, done in https://github.com/grpc/grpc-java/pull/11584/commits/f7574613753dea7ae4c178cc33336a858c0571c0.